Description: Handle possible X error when creating OpenGL context
 Without this higan just crashes when the driver does
 not support OpenGL 3.2 and context creation fails. 
Author: Tobias Hansen <thansen@debian.org>

--- a/ruby/video/glx.cpp
+++ b/ruby/video/glx.cpp
@@ -1,10 +1,19 @@
 #include "opengl/opengl.hpp"
+#include "emulator/emulator.hpp"
+#include "target-ethos/interface/interface.hpp"
 
 #define GLX_CONTEXT_MAJOR_VERSION_ARB 0x2091
 #define GLX_CONTEXT_MINOR_VERSION_ARB 0x2092
 
 namespace ruby {
 
+static bool ctxErrorOccurred = false;
+static int ctxErrorHandler( Display *dpy, XErrorEvent *ev )
+{
+    ctxErrorOccurred = true;
+    return 0;
+}
+
 struct pVideoGLX : OpenGL {
   GLXContext (*glXCreateContextAttribs)(Display*, GLXFBConfig, GLXContext, int, const int*) = nullptr;
   int (*glXSwapInterval)(int) = nullptr;
@@ -171,6 +180,10 @@
       XNextEvent(display, &event);
     }
 
+    // Install X error handler to prevent a crash in case context creation fails.
+    ctxErrorOccurred = false;
+    int (*oldHandler)(Display*, XErrorEvent*) = XSetErrorHandler(&ctxErrorHandler);
+
     glxcontext = glXCreateContext(display, vi, /* sharelist = */ 0, /* direct = */ GL_TRUE);
     glXMakeCurrent(display, glxwindow = xwindow, glxcontext);
 
@@ -185,13 +198,24 @@
         None
       };
       GLXContext context = glXCreateContextAttribs(display, fbConfig[0], nullptr, true, attributes);
-      if(context) {
+
+      // Sync to ensure any errors generated are processed.
+      XSync( display, False );
+      if ( !ctxErrorOccurred && context ) {
         glXMakeCurrent(display, 0, nullptr);
         glXDestroyContext(display, glxcontext);
         glXMakeCurrent(display, glxwindow, glxcontext = context);
       }
     }
 
+    XSync( display, False );
+    // Restore the original error handler.
+    XSetErrorHandler( oldHandler );
+    if ( ctxErrorOccurred ) {
+        interface->notify("Error: Creation of an OpenGL 3.2 context failed. Select another video driver on the Advanced Configuration tab and restart higan.\n");
+        return false;
+    }
+
     if(glXSwapInterval) {
       glXSwapInterval(settings.synchronize);
     }
